common/src/main/java/com/thoughtworks/go/remote/work/BuildWork.java

```diff
@@ -41,55 +41,79 @@
+    private transient PluginRequestProcessorRegistry pluginRequestProcessorRegistry;
 
     public BuildWork(BuildAssignment assignment, String consoleLogCharset) {
         this.assignment = assignment;
         this.consoleLogCharset = consoleLogCharset;
     }
 
     private void initialize(AgentWorkContext agentWorkContext) {
-        JobIdentifier jobIdentifier = assignment.getJobIdentifier();
+        final JobIdentifier jobIdentifier = assignment.getJobIdentifier();
+        final AgentIdentifier agentIdentifier = agentWorkContext.getAgentIdentifier();
+        final AgentRuntimeInfo agentRuntimeInfo = agentWorkContext.getAgentRuntimeInfo();
+        pluginRequestProcessorRegistry = agentWorkContext.getPluginRequestProcessorRegistry();
 
+        agentRuntimeInfo.busy(new AgentBuildingInfo(jobIdentifier.buildLocatorForDisplay(), jobIdentifier.buildLocator()));
         this.timeProvider = new TimeProvider();
-        agentWorkContext.getAgentRuntimeInfo().busy(new AgentBuildingInfo(jobIdentifier.buildLocatorForDisplay(), jobIdentifier.buildLocator()));
         this.workingDirectory = assignment.getWorkingDirectory();
         this.materialRevisions = assignment.materialRevisions();
-        this.goPublisher = new DefaultGoPublisher(agentWorkContext.getArtifactsManipulator(), jobIdentifier, agentWorkContext.getRepositoryRemote(), agentWorkContext.getAgentRuntimeInfo(), consoleLogCharset);
-        this.artifactsPublisher = new ArtifactsPublisher(goPublisher, agentWorkContext.getArtifactExtension(), assignment.getArtifactStores(), agentWorkContext.getPluginRequestProcessorRegistry(), workingDirectory);
-        this.builders = new Builders(assignment.getBuilders(), goPublisher, agentWorkContext.getTaskExtension(), agentWorkContext.getArtifactExtension(), agentWorkContext.getPluginRequestProcessorRegistry());
+        this.goPublisher = new DefaultGoPublisher(agentWorkContext.getArtifactsManipulator(), jobIdentifier, agentWorkContext.getRepositoryRemote(), agentRuntimeInfo, consoleLogCharset);
+        this.artifactsPublisher = new ArtifactsPublisher(goPublisher, agentWorkContext.getArtifactExtension(), assignment.getArtifactStores(), workingDirectory);
+        this.builders = new Builders(assignment.getBuilders(), goPublisher, agentWorkContext.getTaskExtension(), agentWorkContext.getArtifactExtension());
+
+        this.materialAgentFactory = new MaterialAgentFactory(workingDirectory, agentIdentifier, agentWorkContext.getPackageRepositoryExtension(), agentWorkContext.getScmExtension());
     }
 
     public void doWork(EnvironmentVariableContext environmentVariableContext, AgentWorkContext agentWorkContext) {
         initialize(agentWorkContext);
         try {
-            JobResult result = build(environmentVariableContext, agentWorkContext.getAgentIdentifier(), agentWorkContext.getPackageRepositoryExtension(), agentWorkContext.getScmExtension());
+            registerConsoleLogProcessors();
```

Comment:
Shouldn't `registerConsoleLogProcessors` and `removeConsoleLogProcessors` happen for the `BuildSession` workflow too?

Reply:
Why not. I will update the BuildSession workflow.

Reply:
This would be handled in https://github.com/gocd/gocd/issues/4401